src/utils/userLocation.ts

```diff
@@ -1,81 +1,129 @@
+};
+
+export const updateCurrentLocationData = ({
+  userId,
+  profileLocationData,
+}: UpdateCurrentLocationDataProps) => {
+  const [locationName] = Object.keys(profileLocationData);
+
+  updateLocationData({
+    userId,
+    newLocationData: { [locationName]: getCurrentTimeInMilliseconds() },
   });
 };
 
-export const enterVenue = (venueId: string) => openUrl(venueInsideUrl(venueId));
+export const clearLocationData = (userId: string) => {
+  updateLocationData({ userId, newLocationData: {} });
+};
 
-type EnterExternalRoomProps = {
+export type EnterExternalRoomProps = {
   userId: string;
-  room: Room;
+  roomUrl: string;
+  roomId: string;
 };
 
-export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
-  trackLocationEntered({
+export const enterExternalRoom = ({
+  userId,
+  roomId,
+  roomUrl,
+}: EnterExternalRoomProps) => {
+  setLocationData({
     userId,
-    locationName: room.url,
+    locationName: roomId,
   });
 
-  openRoomUrl(room.url);
+  openRoomUrl(roomUrl);
+};
+
+export type UseUpdateTimespentPeriodicallyProps = {
+  locationName: string;
+  userId?: string;
 };
 
-export const useUpdateTimespentPeriodically = (
-  user: UserInfo | undefined,
-  roomName: string
-) => {
-  const shouldUseInterval = user && roomName;
+// @dept I don't think this functinality works correctly, since we only log `internal venues` in this piece of code
+// Could be also beneficial to log external rooms' timespent
+export const useUpdateTimespentPeriodically = ({
+  locationName,
+  userId,
+}: UseUpdateTimespentPeriodicallyProps) => {
+  const shouldUseInterval = userId && locationName;
 
   useInterval(
     () => {
-      // Time spent is currently counted multiple time if multiple tabs are open
-      if (!user || !roomName) return;
+      // @debt time spent is currently counted multiple times if multiple tabs are open
+      if (!userId || !locationName) return;
 
       const firestore = firebase.firestore();
-      const doc = `users/${user.uid}/visits/${roomName}`;
-      const increment = firebase.firestore.FieldValue.increment(
-        LOCATION_INCREMENT_SECONDS
-      );
 
       return firestore
-        .doc(doc)
-        .update({ timeSpent: increment })
+        .collection("users")
+        .doc(userId)
+        .collection("visits")
+        .doc(locationName)
+        .update({
+          timeSpent: firebase.firestore.FieldValue.increment(
+            LOCATION_INCREMENT_SECONDS
+          ),
+        })
         .catch(() => {
-          firestore.doc(doc).set({ timeSpent: LOCATION_INCREMENT_SECONDS });
+          /*
+            NOTE: it's (intended to be) because there was no document to update,
+            and so we are defaulting it to LOCATION_INCREMENT_SECONDS
+            since that's how often this code runs
+          */
+          firestore
+            .collection("users")
+            .doc(userId)
+            .collection("visits")
+            .doc(locationName)
```

Comment:
If you set just this part as a `const` that will give you the 'ref', you can then use it for both the `.update` and the `.set` without having to duplicate the 'path'.
```suggestion
          userLocationRef
```

Reply:
fixed